tests/test_identity.py

```diff
@@ -26,27 +28,764 @@ def register_opts(self, opts, group=None):
+            "keystone_keycloak_backend.identity.KeycloakOpenID"
+        ) as mock_openid_class:
+
+            # Mock token acquisition
+            mock_openid = Mock()
+            mock_openid_class.return_value = mock_openid
+            mock_openid.token.return_value = {"access_token": "test-token"}
+
+            # This should raise an Exception
+            with pytest.raises(Exception):
+                service_account_driver._keycloak_with_retry(mock_operation)
+
+        # Verify NO detailed error logging happened (debug=false)
+        error_calls = [
+            call
+            for call in mock_log.error.call_args_list
+            if "KEYCLOAK_ERROR:" in str(call)
+        ]
+        assert len(error_calls) == 0, "Should not log detailed errors when debug=false"
+
+    def test_attribute_error_no_longer_handled(self, service_account_driver):
+        """Test that AttributeError from python-keycloak 3.x is no longer an issue.
+
+        Since we switched to direct authentication approach, we no longer trigger
+        the problematic code path that caused the AttributeError in python-keycloak 3.x.
+        """
+        # Mock operation that would have triggered AttributeError in old token-based approach
+        mock_operation = Mock(
+            side_effect=AttributeError("'NoneType' object has no attribute 'get'")
+        )
+        mock_operation.__name__ = "get_user"
+
+        # This AttributeError should now be raised normally since we don't handle it
+        # (because we no longer trigger the problematic code path)
+        with pytest.raises(
+            AttributeError, match="'NoneType' object has no attribute 'get'"
+        ):
+            service_account_driver._keycloak_with_retry(mock_operation)
+
+    def test_attribute_error_no_longer_handled_debug(self, debug_driver):
+        """Test that AttributeError is no longer handled even with debug mode.
+
+        Since we switched to direct authentication, we no longer trigger the
+        problematic python-keycloak 3.x code path.
+        """
+        # Mock operation that simulates the old AttributeError
+        original_operation = Mock(
+            side_effect=AttributeError("'NoneType' object has no attribute 'get'")
+        )
+        original_operation.__name__ = "get_user"
+
+        # This should raise AttributeError normally (no special handling)
+        with pytest.raises(
+            AttributeError, match="'NoneType' object has no attribute 'get'"
+        ):
+            debug_driver._keycloak_with_retry(original_operation)
+
+    def test_non_token_attribute_error_passthrough(self, service_account_driver):
+        """Test that non-token AttributeErrors are passed through normally."""
+        # Mock operation that raises unrelated AttributeError
+        mock_operation = Mock(side_effect=AttributeError("Some other attribute error"))
+        mock_operation.__name__ = "get_user"
+
+        # This should raise the AttributeError without retry
```

Comment:
Several tests are skipped with `@pytest.mark.skip(reason="Property mocking complexity - TODO: Fix")`. These tests cover important error handling scenarios (403 errors, retry logic).

These should be fixed before merging to ensure the new retry and error handling logic is properly tested. If the mocking is complex, consider refactoring the code to be more testable (e.g., by extracting the Keycloak client creation into a method that can be easily mocked).